src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+            else:
+                new_key_padding_mask = torch.cat([prev_key_padding_mask, key_padding_mask], dim=1)
+
+        elif key_padding_mask is not None:
+            filler = torch.zeros(
+                batch_size,
+                src_len - key_padding_mask.size(1),
+                dtype=key_padding_mask.dtype,
+                device=key_padding_mask.device,
+            )
+            new_key_padding_mask = torch.cat([filler, key_padding_mask], dim=1)
+        else:
+            new_key_padding_mask = prev_key_padding_mask
+        return new_key_padding_mask
+
+
+class EncoderLayer(nn.Module):
+    def __init__(self, config: ProphetNetConfig):
+        super().__init__()
+        self.embed_dim = config.d_model
+        self.self_attn = SelfAttention(
+            self.embed_dim, config.encoder_attention_heads, dropout=config.attention_dropout,
+        )
+        self.self_attn_layer_norm = LayerNorm(self.embed_dim)
+        self.dropout = config.dropout
+        self.activation_fn = ACT2FN[config.activation_function]
+        self.activation_dropout = config.activation_dropout
+        self.fc1 = nn.Linear(self.embed_dim, config.encoder_ffn_dim)
+        self.fc2 = nn.Linear(config.encoder_ffn_dim, self.embed_dim)
+        self.final_layer_norm = LayerNorm(self.embed_dim)
+
+    def forward(self, x, encoder_padding_mask, output_attentions=False):
+        residual = x
+        x, attn_weights = self.self_attn(
+            query=x, key=x, key_padding_mask=encoder_padding_mask, output_attentions=output_attentions
+        )
+        x = F.dropout(x, p=self.dropout, training=self.training)
+        x = residual + x
+        x = self.self_attn_layer_norm(x)
+        residual = x
+        x = self.activation_fn(self.fc1(x))
+        x = F.dropout(x, p=self.activation_dropout, training=self.training)
+        x = self.fc2(x)
+        x = F.dropout(x, p=self.dropout, training=self.training)
+        x = residual + x
+        x = self.final_layer_norm(x)
+        return x, attn_weights
+
+
+
+class ProphetNetEncoder(nn.Module):
+    """
+    ProphetNet Encoder is same to traditional Transformer Encoder.
+    Transformer encoder consisting of *config.encoder_layers* self attention layers. Each layer
+    is a :class:`EncoderLayer`.
+
+    Args:
+        config: ProphetNetConfig
+    """
+    def __init__(self, config: ProphetNetConfig, embed_tokens):
+        super().__init__()
+
+        self.output_hidden_states = config.output_hidden_states
+
+        self.dropout = config.dropout
+        embed_dim = embed_tokens.embedding_dim
+        self.padding_idx = embed_tokens.padding_idx
+        self.max_source_positions = config.max_position_embeddings
+
+        self.embed_tokens = embed_tokens
+        self.embed_scale = None
+        self.embed_positions = LearnedPositionalEmbedding(config.max_position_embeddings+1+self.padding_idx, embed_dim, self.padding_idx)
+
+        self.layers = nn.ModuleList([])
```

Comment:
why not directly put everything int `nn.ModuleList`? Seems unnecessary to use `.extend(...)`